cpp/oneapi/dal/detail/communicator.hpp

```diff
@@ -0,0 +1,741 @@
+    /// @param dtype      The type of elements in the passed buffers
+    ///
+    /// @return The object to track the progress of the operation
+    spmd_request allgather(const byte_t* send_buf,
+                           std::int64_t send_count,
+                           byte_t* recv_buf,
+                           std::int64_t recv_count,
+                           const data_type& dtype) const;
+
+#ifdef ONEDAL_DATA_PARALLEL
+    /// `allgather` that accepts USM pointers
+    spmd_request allgather(sycl::queue& q,
+                           const byte_t* send_buf,
+                           std::int64_t send_count,
+                           byte_t* recv_buf,
+                           std::int64_t recv_count,
+                           const data_type& dtype) const;
+#endif
+
+    template <typename T, enable_if_primitive_t<T>* = nullptr>
+    spmd_request allgather(const T* send_buf,
+                           std::int64_t send_count,
+                           T* recv_buf,
+                           std::int64_t recv_count) const {
+        return allgather(reinterpret_cast<const byte_t*>(send_buf),
+                         send_count,
+                         reinterpret_cast<byte_t*>(recv_buf),
+                         recv_count,
+                         make_data_type<T>());
+    }
+
+#ifdef ONEDAL_DATA_PARALLEL
+    template <typename T, enable_if_primitive_t<T>* = nullptr>
+    spmd_request allgather(sycl::queue& q,
+                           const T* send_buf,
+                           std::int64_t send_count,
+                           T* recv_buf,
+                           std::int64_t recv_count) const {
+        return allgather(q,
+                         reinterpret_cast<const byte_t*>(send_buf),
+                         send_count,
+                         reinterpret_cast<byte_t*>(recv_buf),
+                         recv_count,
+                         make_data_type<T>());
+    }
+#endif
+
+    template <typename T, enable_if_primitive_t<T>* = nullptr>
+    spmd_request allgather(const array<T>& send, const array<T>& recv) {
+#ifdef ONEDAL_ENABLE_ASSERT
+        check_if_same_send_count(send.get_count(), get_default_root_rank());
+
+        // Check if recv allocated count is enough to store all the received values
+        const std::int64_t min_recv_count = get_min_recv_count(send.get_count());
+        ONEDAL_ASSERT(recv.get_count() >= min_recv_count);
+#endif
+
+        if (send.get_count() == 0) {
+            ONEDAL_ASSERT(recv.get_count() == 0);
+            return spmd_request{};
+        }
+
+        ONEDAL_ASSERT(send.get_count() > 0);
+        ONEDAL_ASSERT(recv.has_mutable_data());
+
+        spmd_request request;
+
+        __ONEDAL_IF_QUEUE__(send.get_queue(), {
+            auto q = send.get_queue().value();
+
+            ONEDAL_ASSERT(recv.get_queue().has_value());
+            ONEDAL_ASSERT(recv.get_queue().value() == q);
```

Comment:
could it be another queue sharing the same context?

Reply:
Ok, will check contexts

Reply:
Done, 9a95578